src/Compilers/VisualBasic/Portable/Parser/ParseQuery.vb

```diff
@@ -711,12 +702,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
                 Predicate = New CodeAnalysis.Syntax.InternalSyntax.SeparatedSyntaxList(Of JoinConditionSyntax)(missingEq)
             End If
 
-            Dim joinVarList = New CodeAnalysis.Syntax.InternalSyntax.SeparatedSyntaxList(Of CollectionRangeVariableSyntax)(joinVariable)
-            Dim moreJoins = moreJoinsBuilder.ToList()
-            _pool.Free(moreJoinsBuilder)
+            Dim joinVarList = _pool.AllocateSeparated(Of CollectionRangeVariableSyntax)
+            joinVarList.Add(joinVariable)
+            Dim moreJoins = _pool.ToListAndFree(moreJoinsBuilder)
 
             If groupKw Is Nothing Then
-                Return SyntaxFactory.SimpleJoinClause(joinKw, joinVarList, moreJoins, onKw, Predicate)
+                Return SyntaxFactory.SimpleJoinClause(joinKw, _pool.ToListAndFree(joinVarList), moreJoins, onKw, Predicate)
```

Comment:
the list seems can be re-used.

Reply:
different lists `moreJoinsBuilder` and `joinVarList`

Reply:
@heejaechang  I've made them extension methods, so that it better matches meaning.

Reply:
Also they are on different function termination paths so can be reused.